lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +619,233 @@ type SelectHopHintsCfg struct {
 	// GetAlias allows the peer's alias SCID to be retrieved for private
 	// option_scid_alias channels.
 	GetAlias func(lnwire.ChannelID) (lnwire.ShortChannelID, error)
+
+	// FetchAllChannels retrieves all open channels currently stored
+	// within the database.
+	FetchAllChannels func() ([]*channeldb.OpenChannel, error)
+
+	// IsChannelActive checks whether the channel identified by the provided
+	// ChannelID is considered active.
+	IsChannelActive func(chanID lnwire.ChannelID) bool
+
+	// MaxHopHints is the maximum number of hop hints we are interested in.
+	MaxHopHints int
 }
 
-func newSelectHopHintsCfg(invoicesCfg *AddInvoiceConfig) *SelectHopHintsCfg {
+func newSelectHopHintsCfg(invoicesCfg *AddInvoiceConfig,
+	maxHopHints int) *SelectHopHintsCfg {
+
 	return &SelectHopHintsCfg{
+		FetchAllChannels:      invoicesCfg.ChanDB.FetchAllChannels,
+		IsChannelActive:       invoicesCfg.IsChannelActive,
 		IsPublicNode:          invoicesCfg.Graph.IsPublicNode,
 		FetchChannelEdgesByID: invoicesCfg.Graph.FetchChannelEdgesByID,
 		GetAlias:              invoicesCfg.GetAlias,
+		MaxHopHints:           maxHopHints,
 	}
 }
 
 // sufficientHints checks whether we have sufficient hop hints, based on the
-// following criteria:
-//   - Hop hint count: limit to a set number of hop hints, regardless of whether
-//     we've reached our invoice amount or not.
-//   - Total incoming capacity: limit to our invoice amount * scaling factor to
-//     allow for some of our links going offline.
+// any of the following criteria:
+//   - Hop hint count: the number of hints have reach our max target.
+//   - Total incoming capacity: the sum of the remote balance amount in the
+//     hints is bigger of equal than our target (currently twice the invoice
+//     amount)
 //
 // We limit our number of hop hints like this to keep our invoice size down,
 // and to avoid leaking all our private channels when we don't need to.
-func sufficientHints(numHints, maxHints, scalingFactor int, amount,
-	totalHintAmount lnwire.MilliSatoshi) bool {
+func sufficientHints(nHintsLeft int, currentAmount,
+	targetAmount lnwire.MilliSatoshi) bool {
 
-	if numHints >= maxHints {
-		log.Debug("Reached maximum number of hop hints")
+	if nHintsLeft <= 0 {
+		log.Debugf("Reached targeted number of hop hints")
 		return true
 	}
 
-	requiredAmount := amount * lnwire.MilliSatoshi(scalingFactor)
```

Comment:
What happened to the scaling factor here?

Reply:
`PopulateHopHints` has it for calculating `targetBandwidth := amtMSat * hopHintFactor
`